src/NHibernate/Linq/Visitors/NullableExpressionDetector.cs

```diff
@@ -0,0 +1,288 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using NHibernate.Engine;
+using NHibernate.Linq.Clauses;
+using NHibernate.Linq.Expressions;
+using NHibernate.Linq.Functions;
+using NHibernate.Util;
+using Remotion.Linq.Clauses;
+using Remotion.Linq.Clauses.Expressions;
+using Remotion.Linq.Clauses.ResultOperators;
+
+namespace NHibernate.Linq.Visitors
+{
+	internal class NullableExpressionDetector
+	{
+		private static readonly HashSet<System.Type> NotNullOperators = new HashSet<System.Type>
+		{
+			typeof(AllResultOperator),
+			typeof(AnyResultOperator),
+			typeof(ContainsResultOperator),
+			typeof(CountResultOperator),
+			typeof(LongCountResultOperator)
+		};
+
+		private readonly Dictionary<BinaryExpression, List<MemberExpression>> _equalityNotNullMembers =
+			new Dictionary<BinaryExpression, List<MemberExpression>>();
+		private readonly ISessionFactoryImplementor _sessionFactory;
+		private readonly ILinqToHqlGeneratorsRegistry _functionRegistry;
+
+		public NullableExpressionDetector(ISessionFactoryImplementor sessionFactory, ILinqToHqlGeneratorsRegistry functionRegistry)
+		{
+			_sessionFactory = sessionFactory;
+			_functionRegistry = functionRegistry;
+		}
+
+		public void SearchForNotNullMemberChecks(BinaryExpression expression)
+		{
+			// Check for a member not null check that has a not equals expression
+			// Example: o.Status != null && o.Status != "New"
+			// Example: (o.Status != null && o.OldStatus != null) && (o.Status != o.OldStatus)
+			// Example: (o.Status != null && o.OldStatus != null) && (o.Status == o.OldStatus)
+			if (
+				expression.NodeType != ExpressionType.AndAlso ||
```

Comment:
I would include the `ExpressionType.And` case too here. I do not see any value in not optimizing the generated query for non-short-circuiting `&`. I think we should optimize all we can as long as the queries yielded result sets stay the same.

Reply:
> I would include the ExpressionType.And case too here

`ExpressionType.And` is now added, but I wasn't able to add any tests as it seems that `QueryTranslatorImpl` does not support such cases.

Reply:
You mean, using simple `&` is not supported by the Linq provider? (Or maybe ambiguous with binary "and".) I have never checked that indeed.

Reply:
> You mean, using simple & is not supported by the Linq provider?

It is supported but not when the left or the right expression is a binary expression. For example:
```
var list =  session.Query<Product>().Where(o => o.Discontinued & o.Discontinued).ToList(); // Works
var list2 =  session.Query<Product>().Where(o => o.Discontinued & o.Name != null).ToList(); // Throws
```
the first query does work but the second doesn't. It seems like that neither Sql Server 2017 supports it:
`SELECT * FROM Products where Discontinued = 1 & (ProductName != 'test')` fails with error: `Incorrect syntax near '!'.` ([fiddle](http://sqlfiddle.com/#!18/41386/1))

Reply:
I do not get the point with Sql Server. In T-SQL `&` is for "binary and", while `&` in C# has a dual usage : it may be the binary operator or the non-short-circuiting "logical and" (when both side are boolean expressions), which is just `and` in T-SQL. T-SQL does not have a short-circuiting "logical and" operator.
The correct translation for C# `&` when both sides are boolean is the T-SQL `and`, not `&`. So it should be supported I think, but that is not the subject of this PR.

Overall I do not think we need to change anything more in this PR about this, so I am off-topic.

Reply:
> T-SQL does not have a short-circuiting "logical and" operator.

This is because this concept does not apply to the SQL server.

Reply:
You are right, I got confused. Indeed, this should be covered in a separate PR.

Reply:
>>T-SQL does not have a short-circuiting "logical and" operator.
>
> This is because this concept does not apply to the SQL server.

And to SQL at all I think indeed. It is more an iterative concept than a thing for data manipulation.